File: airesources/JavaScript/networking.js
'use strict'
const hlt = require('./hlt.js');
const readline = require('readline');
var rl = readline.createInterface({
	input: process.stdin,
	output: process.stdout,
	terminal: false
});
var lines = []
rl.on('line', function(line){
	lines.push(line);
})
var _productions = []
var _width = -1
var _height = -1

function serializeMoveSet(moves) {
	var returnString = "";
	for (var move in move) {
		returnString += move.loc.x.toString() + " " + move.loc.y.toString() + " " + move.direction.toString() + " ";
	}
	return returnString;
}
function deserializeMapSize(inputString) {
	splitString = inputString.split(" ");
	_width = parseInt(splitString.shift());
	_height = parseInt(splitString.shift());
}
function deserializeProductions(inputString) {
	splitString = inputString.split(" ");
	for (var a = 0; a < _height; a++) {
		row = [];
		for (var b = 0; b < _width; b++) {
			row.push(parseInt(splitString.shift()));
		}
		_productions.push(row);
	}
}
function deserializeMap(inputString) {
	splitString = inputString.split(" ");
	m = new hlt.GameMap(_width, _height);
	y = 0;
	x = 0;
	counter = 0;
	owner = 0;
	while (y != m.height) {
		counter = parseInt(splitString.shift());
		owner = parseInt(splitString.shift());
		for (var a = 0; a < counter; a++) {
			m.contents[y][x].owner = owner;
			x+=1;
			if (x==m.width) {
				x=0;
				y+=1;
			}
		}
	}
	for (var a = 0; a < _height; a++) {
		for (var b = 0; b < _width; b++) {
			m.contents[a][b].strength = parseInt(splitString.shift());
			m.contents[a][b].production = _productions[a][b];
		}
	}
	return m;
}
function sendString(toBeSent) {
	toBeSent += '\n';
	process.stdout.write(toBeSent);
}
function getString() {
	while (true) {
		if (lines.len > 0) {
			return lines.shift();
		}
	}
}
/*function hiddenGetInit() { // for sync reasons
	// var playerTag, m;
	// return getString().then(function (a) {
	// 	playerTag = parseInt(a);
	// 	return getString().then(function (a) {
	// 		deserializeMapSize(a);
	// 		return getString().then(function (a) {
	// 			deserializeProductions(a);
	// 			return getString().then(function (a) {
	// 				m = deserializeMap(d);
	// 				return [playerTag, d];
	// 			})
	// 		})
	// 	})
	// })
	var a = getString(),
		b = getString(),
		c = getString(),
		d = getString(),
	playerTag = a,
	a = deserializeMapSize(b),
	b = deserializeProductions(c),
	m = deserializeMap(d);
	return [playerTag, m]
}*/
function getInit() {
	var playerTag = parseInt(getString());
	deserializeMapSize(getString());
	deserializeProductions(getString());
	var m = deserializeMap(getString())
	return [playerTag, m]

	/* // Attempt at a better sync thing
	return new Promise(function(resolve, reject) {
		var playerTag, m;
		resolve(
			getString().then(function (a) {
				playerTag = parseInt(a);
				return getString().then(function (a) {
					deserializeMapSize(a);
					return getString().then(function (a) {
						deserializeProductions(a);
						return getString().then(function (a) {
							m = deserializeMap(d);
							return [playerTag, m];
						})
					})
				})
			}))
	}); */

	//return setTimeout(hiddenGetInit,200);
}
function sendInit(name) {
	sendString(name);
}
function hiddenGetFrame() {
	return getString().then(function (a) {
		return deserializeMap(a);
	});
}
function getFrame() {
	return new Promise(function(resolve, reject) {
		resolve( setTimeout(hiddenGetFrame,200));
	});
}
function sendFrame(moves) {
	sendString(serializeMoveSet(moves));
}
module.exports = {getInit, sendInit, getFrame, sendFrame}

File: airesources/Elixir/lib/networking.ex
defmodule Networking do
    defp get_string do
        IO.gets("")
        |> String.strip
    end

    defp send_string(string) do
        IO.puts string
    end

    defp int_split(string) do
        String.split(string)
        |> Enum.map(fn(s) ->
            Integer.parse(s)
            |> elem(0)
        end)
    end
    
    defp serialize_move_set(moves) do
        Enum.map(moves, fn({l, d}) ->
            "#{l.x} #{l.y} #{d}"
        end)
        |> Enum.join(" ")
    end

    defp deserialize_game_map_size(string) do
        int_split(string)
        |> List.to_tuple
    end

    defp deserialize_productions(string, width, height) do
        locations = GameMap.locations(width, height)
        productions = int_split(string)
        map = Enum.zip(locations, productions)
        |> Map.new(fn({location, production}) ->
            {location, %Site{production: production}}
        end)
        %GameMap{width: width, height: height, map: map}
    end

    defp deserialize_game_map(string, game_map) do
        split = int_split(string)
        idx = length(split) - game_map.width * game_map.height
        {owners, strengths} = Enum.split(split, idx)
        owners = Enum.chunk(owners, 2)
        |> Enum.flat_map(fn([count, tag]) ->
            List.duplicate(tag, count)
        end)
        locations = GameMap.locations(game_map)
        map = List.zip([locations, owners, strengths])
        |> Enum.reduce(game_map.map, fn({location, owner, strength}, acc) ->
            Map.update!(acc, location, fn(site) ->
                %Site{site | owner: owner, strength: strength}
            end)
        end)
        %GameMap{game_map | map: map}
    end

    def get_init do
        {player_tag, _} = Integer.parse(get_string)
        {width, height} = deserialize_game_map_size(get_string)
        game_map = deserialize_productions(get_string, width, height)
        game_map = deserialize_game_map(get_string, game_map)
        {player_tag, game_map}
    end

    def send_init(name) do
        send_string(name)
    end

    def get_frame(game_map) do
        deserialize_game_map(get_string, game_map)
    end

    def send_frame(moves) do
        send_string(serialize_move_set(moves))
    end
end